File: go/space-age/space_age.go
package space

// Planet represent a planet name
type Planet string

const earthSecPerYear = 31557600

var fromPlanetRatio = map[Planet]float64{
	"Mercury": 0.2408467 * earthSecPerYear,
	"Venus":   0.61519726 * earthSecPerYear,
	"Earth":   1 * earthSecPerYear,
	"Mars":    1.8808158 * earthSecPerYear,
	"Jupiter": 11.862615 * earthSecPerYear,
	"Saturn":  29.447498 * earthSecPerYear,
	"Uranus":  84.016846 * earthSecPerYear,
	"Neptune": 164.79132 * earthSecPerYear,
}

// Age converts an age on earth to an age on the requested planet
func Age(seconds float64, p Planet) float64 {
	return seconds / fromPlanetRatio[p]
}

File: ruby/space-age/space_age.rb
class SpaceAge
  EARTH_SECOND_PER_YEAR = 31_557_600.to_f
  PLANETS_TO_EARTH_ORBITAL_PERIODS = {
    mercury: 0.2408467,
    venus: 0.61519726,
    earth: 1,
    mars: 1.8808158,
    jupiter: 11.862615,
    saturn: 29.447498,
    uranus: 84.016846,
    neptune: 164.79132
  }.freeze

  private_constant :PLANETS_TO_EARTH_ORBITAL_PERIODS, :EARTH_SECOND_PER_YEAR

  PLANETS_TO_EARTH_ORBITAL_PERIODS.each do |planet, period|
    define_method("on_#{planet}".to_sym) { convert(period) }
  end

  private

  def initialize(seconds)
    @seconds = seconds
  end

  def convert(period)
    (@seconds / EARTH_SECOND_PER_YEAR) / period
  end
end
